File: app/src/main/java/com/originb/inkwisenote2/modules/noterelation/NoteRelationEventListener.java
package com.originb.inkwisenote2.modules.noterelation;

import com.originb.inkwisenote2.modules.backgroundjobs.BackgroundOps;
import com.originb.inkwisenote2.modules.backgroundjobs.Events;
import com.originb.inkwisenote2.modules.repositories.NoteRelationRepository;
import com.originb.inkwisenote2.modules.repositories.SmartNotebook;
import org.greenrobot.eventbus.EventBus;
import org.greenrobot.eventbus.Subscribe;
import org.greenrobot.eventbus.ThreadMode;
import org.koin.java.KoinJavaComponent;

public class NoteRelationEventListener {
    private NoteRelationRepository noteRelationRepository;

    public NoteRelationEventListener(NoteRelationRepository noteRelationRepository) {
        // Inject NoteRelationRepository via Koin
        this.noteRelationRepository = noteRelationRepository;
        EventBus.getDefault().register(this);
    }

    @Subscribe(threadMode = ThreadMode.BACKGROUND)
    public void onDataUpdateEvent(Events.NotebookDeleted notebookToDelete) {
        BackgroundOps.execute(() -> {
            SmartNotebook smartNotebook = notebookToDelete.smartNotebook;
            smartNotebook.atomicNotes.forEach(note ->
                    noteRelationRepository.deleteNoteRelationData(note)
            );
        });
    }

    @Subscribe(threadMode = ThreadMode.BACKGROUND)
    public void onNoteDelete(Events.NoteDeleted noteDeleted) {
        BackgroundOps.execute(() -> BackgroundOps.execute(() ->
                noteRelationRepository.deleteNoteRelationData(noteDeleted.atomicNote)
        ));
    }
}

File: app/src/main/java/com/originb/inkwisenote2/modules/noterelation/NoteRelationEventListener.kt
package com.originb.inkwisenote2.modules.noterelation

import com.originb.inkwisenote2.modules.backgroundjobs.BackgroundOps.Companion.execute
import com.originb.inkwisenote2.modules.backgroundjobs.Events.NoteDeleted
import com.originb.inkwisenote2.modules.backgroundjobs.Events.NotebookDeleted
import com.originb.inkwisenote2.modules.repositories.NoteRelationRepository
import com.originb.inkwisenote2.modules.smartnotes.data.AtomicNoteEntity
import org.greenrobot.eventbus.EventBus
import org.greenrobot.eventbus.Subscribe
import org.greenrobot.eventbus.ThreadMode
import java.util.function.Consumer

class NoteRelationEventListener(private val noteRelationRepository: NoteRelationRepository) {
    init {
        // Inject NoteRelationRepository via Koin
        EventBus.getDefault().register(this)
    }

    @Subscribe(threadMode = ThreadMode.BACKGROUND)
    fun onDataUpdateEvent(notebookToDelete: NotebookDeleted) {
        execute {
            val smartNotebook = notebookToDelete.smartNotebook
            smartNotebook.atomicNotes.forEach(Consumer { note: AtomicNoteEntity ->
                noteRelationRepository.deleteNoteRelationData(
                    note
                )
            }
            )
        }
    }

    @Subscribe(threadMode = ThreadMode.BACKGROUND)
    fun onNoteDelete(noteDeleted: NoteDeleted) {
        execute(Runnable {
            execute(Runnable { noteRelationRepository.deleteNoteRelationData(noteDeleted.atomicNote) }
            )
        })
    }
}
